solver/pb/ops.proto

```diff
@@ -211,4 +199,67 @@ message Definition {
 message HostIP {
 	string Host = 1;
 	string IP = 2;
+}
+
+message FileOp {
+	repeated FileAction actions = 2;
+}
+
+message FileAction {
+	int64 input = 1 [(gogoproto.customtype) = "InputIndex", (gogoproto.nullable) = false]; // could be real input or target (target index + max input index)
+	int64 secondaryInput = 2 [(gogoproto.customtype) = "InputIndex", (gogoproto.nullable) = false]; // --//--
+  int64 output = 3 [(gogoproto.customtype) = "OutputIndex", (gogoproto.nullable) = false];
+	oneof action {
+		FileActionCopy copy = 4;
+		FileActionMkFile mkfile = 5;
+		FileActionMkDir mkdir = 6;
+		FileActionRm rm = 7;
+	}
+}
+
+message FileActionCopy {
+	string src = 1;
+	string dest = 2;
+	ChownOpt owner = 4;
+	int32 mode = 5;
+	bool followSymlink = 6;
+	bool dirCopyContents = 7;
+	bool attemptUnpack = 8;
+	bool createDestPath = 9;
+	bool allowWildcard = 10;
+	bool allowEmptyWildcard = 11;
+	int64 timestamp = 12;
+}
+
+message FileActionMkFile {
+	string path = 1;
+	int32 mode = 2;
+	bytes data = 3;
+	ChownOpt owner = 4;
+	int64 timestamp = 5;
+}
+
+message FileActionMkDir {
+	string path = 1;
+	int32 mode = 2;
+	bool makeParents = 3;
+	ChownOpt owner = 4;
+	int64 timestamp = 5;
+}
+
+message FileActionRm {
+	string path = 1;
+	bool allowNotFound = 2;
+	bool allowWildcard = 3;
+}
+
+message ChownOpt {
+	UserOpt user = 1;
+	UserOpt group = 2;
+}
+
+message UserOpt {
+	string name = 1;
+	int64 input = 2 [(gogoproto.customtype) = "InputIndex", (gogoproto.nullable) = false]; // input that contains /etc/passwd if using a name
+	uint32 id = 3;
```

Comment:
Should `name` and `id` be within a `oneof`? What happens if they disagree with the `/etc/passwd` ref'd by `input`?

or perhaps `name`+`input` in one half of a `oneof` and `id` in the other? e.g.
```protobuf
message NamedUserOpt {
	string name = 1;
	int64 input = 2 [(gogoproto.customtype) = "InputIndex"]; // input that contains /etc/passwd
}

message UserOpt {
	oneof user {
		NamedUserOpt byname  = 1;
		uint32 byid  = 2;
	}
}
```
(nb: `input` is not `nullable` here)

Reply:
Yes, this is the correct logic. Left it out as the type was very simple but can do with `oneof`.